src/lua/uri.lua

```diff
@@ -218,6 +253,50 @@ local function values(...)
     return {...}
 end
 
+local char_to_hex = function(c)
+    return string.format("%%%02X", string.byte(c))
+end
+
+-- Encodes a string into its escaped hexadecimal representation.
+local function str_escape(buf, opts)
+    if type(buf) ~= "string" then
+        error("Usage: str_escape(string, [opts])")
+    end
+    opts = opts or RFC3986
+    opts.unreserved = opts.unreserved or RFC3986.unreserved
+    opts.plus = opts.plus or RFC3986.plus
+
+    local url = buf
+    local pattern = ("([%s])"):format(opts.unreserved)
+    url = url:gsub("\n", "\r\n")
```

Comment:
Is it for form data? I doubt that we should do that for query parameters. At least it breaks the `uri.unescape(uri.escape(src)) == src` invariant for any string that contains a line feed character.

Reply:
It should resolve the problem, thanks.

I propose to think about this escaping/unescaping as about something very similar to a data transmitting mechanism. A network card performs encoding from a binary data into a physical signal and this (or another) netword card able to unambiguously reconstruct the original data from this signal. Similarly, we want to transmit a data to another machine as part of a URI: in a path segment, in a query parameter and so on. We have some binary data and a receiver should get it back on the other end.

Reply:
If we actually need it, let's do it for only data forms and consistently for escaping and unescaping.

However it is unclear what exactly is meant by—

> Line breaks are represented as "CR LF" pairs (i.e., `%0D%0A`).

—in the W3C 'standard'. `'foo\nbar'` has no line break, since it has no `'\r\n'` inside? So it should be encoded as `'foo%0Abar'`. Or `'\n'` is the line break, but it should be encoded as `%0D0A`? How to differentiate `'\n'` in a source data from `'\r\n'` after the encoding? Or it assumes some particular interpretation like using in CGI? This part of the document does not answer to those questions. I can interpret it differently without formal inconsistency with the wording.

There is the WHATWG group's document, which looks more normative: https://url.spec.whatwg.org/#application/x-www-form-urlencoded. It has no ambiguity and it doesn't suggest to do any irreversible replacements.

Anyway, it is nice that you highlighted this problem. I propose to look at popular implementations and discuss/decide.